### PR TITLE
feat(connector): [PAYBOX] Connector Integration

### PR DESCRIPTION
**Pull request by aadityaguptaa**
## Description
Added Paybox Connector implementation

### Additional Changes
- [ ] This PR modifies the API contract
- [ ] This PR modifies application configuration/environment variables

### Diffchecker Results
https://github.com/juspay/hyperswitch-cloud/issues/12741
https://github.com/juspay/hyperswitch-cloud/issues/12743
https://github.com/juspay/hyperswitch-cloud/issues/12745
https://github.com/juspay/hyperswitch-cloud/issues/12746

### CODE REVIEW

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 1]
```diff
@@ -0,0 +1,1085 @@
...
+    #[serde(rename = "Refusé")]
+    Rejected,
+}
+
+impl From<PayboxStatus> for AttemptStatus {
+    fn from(item: PayboxStatus) -> Self {
+        match item {
+            PayboxStatus::Cancelled => Self::Voided,
+            PayboxStatus::Authorised => Self::Authorized,
+            PayboxStatus::Captured | PayboxStatus::Refunded => Self::Charged,
```
Comment (iemyashasvi) [inline backend/connector-integration/src/connectors/paybox/transformers.rs]: why is refunded mapped to Charged

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 2]
```diff
@@ -0,0 +1,1085 @@
...
+        .as_millis()
+        .to_string();
+
+    timestamp
+        .get(4..)
+        .map(|s| s.to_string())
+        .ok_or(errors::ConnectorError::ParsingFailed.into())
+}
+
+fn generate_date_time() -> CustomResult<String, errors::ConnectorError> {
```
Comment (iemyashasvi) [inline backend/connector-integration/src/connectors/paybox/transformers.rs:153]: can we use a function from utils for this

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 3]
```diff
@@ -0,0 +1,1085 @@
...
+    #[serde(rename = "RANG")]
+    pub rank: Secret<String>,
+    #[serde(rename = "CLE")]
+    pub key: Secret<String>,
+    #[serde(rename = "NUMQUESTION")]
+    pub paybox_request_number: String,
+    #[serde(rename = "MONTANT")]
+    pub amount: MinorUnit,
+    #[serde(rename = "DEVISE")]
+    pub currency: String,
```
Comment (iemyashasvi) [inline backend/connector-integration/src/connectors/paybox/transformers.rs]: can we use an enum for currency

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 4]
```diff
@@ -0,0 +1,1085 @@
...
+    #[serde(rename = "REFERENCE")]
+    pub reference: String,
+    #[serde(rename = "DATEQ")]
+    pub date: String,
```
Comment (iemyashasvi) [inline backend/connector-integration/src/connectors/paybox/transformers.rs:513]: we can use enum here

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 5]
```diff
@@ -0,0 +1,1085 @@
...
+        item: ResponseRouterData<PayboxCaptureResponse, Self>,
+    ) -> Result<Self, Self::Error> {
+        if item.response.response_code == SUCCESS_CODE {
+            // Create connector_metadata with NUMTRANS
+            let connector_metadata = serde_json::json!(PayboxMeta {
+                connector_request_id: item.response.transaction_number.clone()
+            });
+
+            // Manually set connector_meta_data
+            let connector_meta_data = Secret::new(connector_metadata.clone());
```
Comment (iemyashasvi) [inline backend/connector-integration/src/connectors/paybox/transformers.rs]: why are we manually setting it ?

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 6]
```diff
@@ -0,0 +1,1085 @@
...
+    for RouterDataV2<Refund, RefundFlowData, RefundsData, RefundsResponseData>
+{
+    type Error = error_stack::Report<errors::ConnectorError>;
+
+    fn try_from(item: ResponseRouterData<PayboxRefundResponse, Self>) -> Result<Self, Self::Error> {
+        if item.response.response_code == SUCCESS_CODE {
+            Ok(Self {
+                response: Ok(RefundsResponseData {
+                    connector_refund_id: item.response.paybox_order_id.clone(),
+                    refund_status: RefundStatus::Pending,
```
Comment (iemyashasvi) [inline backend/connector-integration/src/connectors/paybox/transformers.rs]: do we map refund to pending in all the cases ?

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 7]
```diff
@@ -0,0 +1,1085 @@
...
+    #[serde(rename = "NUMQUESTION")]
+    pub paybox_request_number: String,
+    #[serde(rename = "DATEQ")]
+    pub date: String,
+    #[serde(rename = "NUMTRANS")]
+    pub transaction_number: String,
+    #[serde(rename = "NUMAPPEL")]
+    pub paybox_order_id: String,
+}
+
```
Comment (iemyashasvi) [inline backend/connector-integration/src/connectors/paybox/transformers.rs]: see if this struct can be reused ?

File: config/production.toml [hunk 8]
```diff
@@ -21,6 +21,7 @@ bypass_proxy_urls = ["localhost", "local"]
 mitm_proxy_enabled = false
 
 [connectors]
+paybox.base_url = "https://preprod-ppps.paybox.com/PPPS.php"
```
Comment (iemyashasvi) [inline config/production.toml]: is this the production URL ?

**Review by iemyashasvi — CHANGES_REQUESTED** [review]

File: backend/connector-integration/src/connectors/paybox/transformers.rs (hunk 2 above)
Comment (aadityaguptaa) [reply backend/connector-integration/src/connectors/paybox/transformers.rs:153]: I'm using format_date from utils only

use common_utils::{
    date_time::{format_date, now, DateFormat},
    errors::CustomResult,
    types::MinorUnit,
};

File: backend/connector-integration/src/connectors/paybox/transformers.rs (hunk 3 above)
Comment (aadityaguptaa) [reply backend/connector-integration/src/connectors/paybox/transformers.rs]: changed it to use Currency

File: backend/connector-integration/src/connectors/paybox/transformers.rs (hunk 5 above)
Comment (aadityaguptaa) [reply backend/connector-integration/src/connectors/paybox/transformers.rs]: Removed it

File: backend/connector-integration/src/connectors/paybox/transformers.rs (hunk 6 above)
Comment (aadityaguptaa) [reply backend/connector-integration/src/connectors/paybox/transformers.rs]: changed to Success

File: config/production.toml (hunk 8 above)
Comment (aadityaguptaa) [reply config/production.toml]: changed to production url

File: backend/connector-integration/src/connectors/paybox/transformers.rs (hunk 1 above)
Comment (aadityaguptaa) [reply backend/connector-integration/src/connectors/paybox/transformers.rs]: changed it to AutoRefunded

File: backend/connector-integration/src/connectors/paybox/transformers.rs (hunk 7 above)
Comment (aadityaguptaa) [reply backend/connector-integration/src/connectors/paybox/transformers.rs]: changed it to use same struct for both psync and rsync

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 9]
```diff
@@ -0,0 +1,1049 @@
...
+    pub date: String,
+    #[serde(rename = "PORTEUR")]
+    pub card_number: RawCardNumber<T>,
+    #[serde(rename = "DATEVAL")]
+    pub expiration_date: Secret<String>,
+    #[serde(rename = "CVV")]
+    pub cvv: Secret<String>,
+    #[serde(rename = "ACTIVITE")]
+    pub activity: String,
+}
```
Comment (AkshayaFoiger) [inline backend/connector-integration/src/connectors/paybox/transformers.rs:189]: NIT: Can we use `#[serde(rename_all = "UPPERCASE")]` except some special cases

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 10]
```diff
@@ -0,0 +1,1049 @@
...
+    #[serde(rename = "AUTORISATION")]
+    pub authorization: Option<String>,
+    #[serde(rename = "CODEREPONSE")]
+    pub response_code: String,
+    #[serde(rename = "COMMENTAIRE")]
+    pub response_message: String,
+    #[serde(rename = "PORTEUR")]
+    pub carrier_id: Option<Secret<String>>,
+    #[serde(rename = "REFABONNE")]
+    pub customer_id: Option<Secret<String>>,
```
Comment (AkshayaFoiger) [inline backend/connector-integration/src/connectors/paybox/transformers.rs:293]: Same here

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 11]
```diff
@@ -0,0 +1,1049 @@
...
+    #[serde(rename = "AUTORISATION")]
+    pub authorization: Option<String>,
+    #[serde(rename = "CODEREPONSE")]
+    pub response_code: String,
+    #[serde(rename = "COMMENTAIRE")]
+    pub response_message: String,
+    #[serde(rename = "STATUS")]
+    pub status: PayboxStatus,
+}
+
```
Comment (AkshayaFoiger) [inline backend/connector-integration/src/connectors/paybox/transformers.rs:456]: Same here

File: config/production.toml [hunk 12]
```diff
@@ -21,6 +21,7 @@ bypass_proxy_urls = ["localhost", "local"]
 mitm_proxy_enabled = false
 
 [connectors]
+paybox.base_url = "https://api.globalgetnet.com"
```
Comment (AkshayaFoiger) [inline config/production.toml]:
```suggestion
paybox.base_url = "https://ppps.paybox.com/PPPS.php"
```

File: backend/connector-integration/src/connectors/paybox/transformers.rs (hunk 9 above)
Comment (aadityaguptaa) [reply backend/connector-integration/src/connectors/paybox/transformers.rs:189]: I've added #[serde(rename_all = "UPPERCASE")] to all the requested structs

File: config/production.toml (hunk 12 above)
Comment (aadityaguptaa) [reply config/production.toml]: done

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 13]
```diff
@@ -0,0 +1,1045 @@
...
+        >,
+    > for PayboxVoidRequest
+{
+    type Error = error_stack::Report<errors::ConnectorError>;
+
+    fn try_from(
+        item: PayboxRouterData<
+            RouterDataV2<Void, PaymentFlowData, PaymentVoidData, PaymentsResponseData>,
+            T,
+        >,
```
Comment (maverox) [inline backend/connector-integration/src/connectors/paybox/transformers.rs]: 🟡 **Warning** - Status Mapping

**Issue**: Unknown payment status maps to `Failure` instead of `Pending` per UCS convention.

**Current Pattern**:
```rust
_ => AttemptStatus::Failure,
```

**Suggested Fix** (per UCS convention):
```rust
_ => AttemptStatus::Pending, // Unknown statuses should default to Pending
```

**Reasoning**: Unknown statuses may be temporary or new API statuses. Mapping to `Pending` allows the system to sync later and get the actual status, rather than prematurely marking as failed.

File: backend/connector-integration/src/connectors/paybox.rs [hunk 14]
```diff
@@ -0,0 +1,931 @@
...
+    member_functions: {
+        fn preprocess_response_bytes<F, FCD, Req, Res>(
+            &self,
+            _req: &RouterDataV2<F, FCD, Req, Res>,
+            bytes: bytes::Bytes,
+        ) -> CustomResult<bytes::Bytes, errors::ConnectorError> {
+            // Paybox can return responses in two formats:
+            // 1. JSON-wrapped: {"response": "URLENCODED_DATA"}
+            // 2. Direct URL-encoded: URLENCODED_DATA
+
```
Comment (maverox) [inline backend/connector-integration/src/connectors/paybox.rs]: 🟡 **Warning** - Robustness

**Issue**: ISO-8859-15 decoding may silently replace invalid characters without logging.

**Current Code**:
```rust
let (decoded_str, _, _) = encoding_rs::ISO_8859_15.decode(url_encoded_string.as_bytes());
```

**Suggested Improvement**:
```rust
let (decoded_str, _encoding, had_errors) = encoding_rs::ISO_8859_15.decode(url_encoded_string.as_bytes());
if had_errors {
    tracing::warn\!(
        "ISO-8859-15 decoding had replacement characters in Paybox response"
    );
}
```

**Benefit**: Alerts when encoding issues occur, aiding debugging of character encoding problems.

File: backend/connector-integration/src/connectors/paybox.rs [hunk 15]
```diff
@@ -0,0 +1,931 @@
+pub mod transformers;
```
Comment (maverox) [inline backend/connector-integration/src/connectors/paybox.rs]: 🟢 **Suggestion** - Testing

**Issue**: No unit tests detected for the Paybox connector.

**Suggested Action**:
Create `backend/connector-integration/tests/paybox_tests.rs` with tests covering:

1. **Request Transformation Tests**:
   - Authorize request building
   - Capture/Void/Refund request building
   - Amount conversion verification

2. **Response Parsing Tests**:
   - Success response parsing
   - Error response parsing
   - URL-encoded to JSON conversion
   - ISO-8859-15 encoding handling

3. **Status Mapping Tests**:
   - Payment status mapping
   - Refund status mapping
   - Edge cases for unknown statuses

**Example Test Structure**:
```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_authorize_request_building() {
        // Test implementation
    }

    #[test]
    fn test_response_parsing_success() {
        // Test URL-encoded response parsing
    }
}
```

File: backend/connector-integration/src/connectors/paybox/transformers.rs [hunk 16]
```diff
@@ -0,0 +1,1045 @@
...
+    #[serde(rename = "CODEREPONSE")]
+    pub response_code: String,
+    #[serde(rename = "COMMENTAIRE")]
+    pub response_message: String,
+    #[serde(rename = "PORTEUR")]
+    pub carrier_id: Option<Secret<String>>,
+    #[serde(rename = "REFABONNE")]
+    pub customer_id: Option<Secret<String>>,
+}
+
```
Comment (iemyashasvi) [inline backend/connector-integration/src/connectors/paybox/transformers.rs:295]: spell checks are failing

**Review by iemyashasvi — APPROVED** [review]
